src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -113,7 +105,28 @@ Status CoreWorkerDirectTaskSubmitter::SubmitTask(TaskSpecification task_spec) {
       RAY_UNUSED(task_finisher_->PendingTaskFailed(
           task_spec.TaskId(), rpc::ErrorType::TASK_CANCELLED, nullptr));
     }
-  });
+  };
+  if (task_spec.IsActorCreationTask()) {
+    // Synchronously register the actor to GCS server.
+    // Previously, we asynchronously registered the actor after all its dependencies were
+    // resolved. This caused a problem: if the owner of the actor dies before dependencies
+    // are resolved, the actor will never be created. But the actor handle may already be
+    // passed to other workers. In this case, the actor tasks will hang forever.
+    // So we fixed this issue by synchronously registering the actor. If the owner dies
+    // before dependencies are resolved, GCS will notice this and mark the actor as dead.
+    auto status = actor_creator_->AsyncRegisterActor(
```

Comment:
Can we feature flag the asynchrony?

Reply:
I can have a feature flag to turn asyn reg to sync reg. But I'd prefer going the async code path by default and turn it off in case of failures